criu/memfd.c

```diff
@@ -0,0 +1,450 @@
+		goto out;
+
+	if (pb_write_one(img, &mie, PB_MEMFD_INODE))
+		goto out;
+
+	flags = (mie.chunks ? COPY_WITH_CHUNKS : 0) | COPY_REOPEN_FD;
+	ret = copy_fd_to_img(fd, img, mie.size, flags);
+
+out:
+	if (img)
+		close_image(img);
+	return ret;
+}
+
+static struct memfd_inode *dump_unique_memfd_inode(int lfd, const char *name, const struct stat *st)
+{
+	/* Similar to dump_ghost_remap */
+	struct memfd_inode *inode;
+
+	list_for_each_entry(inode, &memfd_inodes, list)
+		if ((inode->dev == st->st_dev) && (inode->ino == st->st_ino))
+			return inode;
+
+	inode = xmalloc(sizeof(*inode));
+	if (inode == NULL)
+		return NULL;
+
+	inode->dev = st->st_dev;
+	inode->ino = st->st_ino;
+	inode->id = memfd_inode_ids++;
+
+	if (dump_memfd_inode(lfd, inode->id, name, st)) {
+		xfree(inode);
+		return NULL;
+	}
+
+	list_add_tail(&inode->list, &memfd_inodes);
+
+	return inode;
+}
+
+static int dump_one_memfd(int lfd, u32 id, const struct fd_parms *p)
+{
+	MemfdFileEntry mfe = MEMFD_FILE_ENTRY__INIT;
+	FileEntry fe = FILE_ENTRY__INIT;
+	struct memfd_inode *inode;
+	struct fd_link _link, *link;
+	const char *name;
+
+	if (!p->link) {
+		if (fill_fdlink(lfd, p, &_link))
+			return -1;
+		link = &_link;
+	} else
+		link = p->link;
+
+	strip_deleted(link);
+	name = &link->name[1+MEMFD_PREFIX_LEN];
+
+	inode = dump_unique_memfd_inode(lfd, name, &p->stat);
+	if (!inode)
+		return -1;
+
+	mfe.id = id;
+	mfe.flags = p->flags;
+	mfe.pos = p->pos;
+	mfe.fown = (FownEntry *)&p->fown;
+	mfe.inode_id = inode->id;
+
+	fe.type = FD_TYPES__MEMFD;
+	fe.id = mfe.id;
+	fe.memfd = &mfe;
+
+	return pb_write_one(img_from_set(glob_imgset, CR_FD_FILES), &fe, PB_FILE);
+}
+
+int dump_one_memfd_cond(int lfd, u32 *id, struct fd_parms *parms)
+{
+	if (fd_id_generate_special(parms, id))
+		return dump_one_memfd(lfd, *id, parms);
+	return 0;
+}
+
+const struct fdtype_ops memfd_dump_ops = {
+	.type		= FD_TYPES__MEMFD,
+	.dump		= dump_one_memfd,
+};
+
+
+/*
+ * Restore only
+ */
+
+struct memfd_info {
+	MemfdFileEntry		*mfe;
+	struct file_desc	d;
+	struct memfd_inode	*inode;
+};
+
+static int memfd_open_inode(struct memfd_inode *inode);
+
+static struct memfd_inode *memfd_alloc_inode(int id)
+{
+	struct memfd_inode *inode;
+
+	list_for_each_entry(inode, &memfd_inodes, list)
+		if (inode->id == id)
+			return inode;
+
+	inode = shmalloc(sizeof(*inode));
+	if (!inode)
+		return NULL;
+
+	inode->id = id;
+	mutex_init(&inode->lock);
+	inode->fdstore_id = -1;
+	inode->pending_seals = 0;
+
+	list_add_tail(&inode->list, &memfd_inodes);
+	return inode;
+}
+
+int apply_memfd_seals(void)
+{
+	/*
+	 * We apply the seals after all the mappings are done because the seal
+	 * F_SEAL_FUTURE_WRITE prevents future write access (added in
+	 * Linux 5.1). Thus we must make sure all writable mappings are opened
+	 * before applying this seal.
+	 */
+
+	struct memfd_inode *inode;
+
+	list_for_each_entry(inode, &memfd_inodes, list) {
+		if (!inode->pending_seals)
+			continue;
+
+		int fd = memfd_open_inode(inode);
```

Comment:
Hm, I've missed that part -> as @avagin said before we always put variables declaration at the start of scope.